Add documentation for IoT logging (#4606)
EnMasseProject/enmasse: documentation/modules/proc-iot-logging.adoc

// Module included in the following assemblies:
//
// assembly-installing-manual-steps.adoc
// assembly-installing-kubernetes.adoc

[id='proc-iot-logging-{context}']

= Configuring logging

If the default logging settings are not sufficient, the following
sections describe ways to configure the logging system.

== Configure the global log levels

The global logging configuration is applied to all
services which do not have any explicit logging configuration.

By default the global log level is `info`.

.Prerequisites

> * link:{BookUrlBase}{BaseProductVersion}{BookNameUrl}#iot-installing-services-{context}[The IoT services are installed].

Author: I don't really know what to do here, and I never really understood how this all works. It seems the section is defined as `[id='iot-installing-services-{context}']` in `proc-iot-installing-services.adoc`. What would be the proper replacement, when not using `{context}`? Should the section ID (and all links to it) be changed as well?

> * link:{BookUrlBase}{BaseProductVersion}{BookNameUrl}#iot-installing-services-{context}[The IoT services are installed].

Reviewer: I think in this case, I can try fixing it in the upstream docs (if the link breaks) once it is merged.

.Procedure

. Edit the `IoTConfig` instance named `default`:
+
[options="nowrap",subs="+quotes,attributes"]
----
{cmdcli} edit iotconfig default
----
. Configure the logging options, save and exit the editor:
+
[source,yaml,options="nowrap"]
----
apiVersion: iot.enmasse.io/v1alpha1
kind: IoTConfig
metadata:
  namespace: enmasse-infra
  name: default
spec:
  logging:
    level: info # <1>
    loggers: # <2>
      io.enmasse: debug # <3>
      io.netty: error # <4>
----
<1> The default global log level. If omitted, `info` will be used.
<2> The section for log channel specific entries.
<3> Lowers the filtering to `debug` level for messages in the channel `io.enmasse`

> <4> Raises the filtering to `error` level for messages in the channel `io.netty`

Reviewer 2: So the log filter level or log level ? It feels one is the inverse of the other.

=== Using the service's ConfigMap resource

In addition to providing the custom configuration using the
`IoTConfig`, it is also possible to put the custom logging
configuration into the services's ConfigMap source.

.Procedure

. Edit the `ConfigMap` instance for the service, for example `iot-http-adapter-config` for the HTTP protocol adapter.
+
[options="nowrap",subs="+quotes,attributes"]
----
{cmdcli} edit cm iot-http-adapter-config
----
. Add the XML based logback configuration in the data section with the key `logback-custom.xml`:
+
[source,yaml,options="nowrap"]
----
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: enmasse-infra
  name: iot-http-adapter-config
data:
  application.yaml: … # <1>
  logback-spring.xml: … # <2>
  logback-custom.xml: | # <3>
    <configuration>
      <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
        <encoder>
          <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
        </encoder>
      </appender>
      <root level="debug">
        <appender-ref ref="STDOUT" />
      </root>
    </configuration>
----
<1> The application specific configuration file. This file is generated

>     by the operator. Any changes will be overwritten.

Reviewer: ```suggestion
    and overwrites any changes.
```

<2> The effective logback configuration, applied by the system. Do not
    change this, as it will be overwritten by the operator.

> <3> The full XML based logback configuration

Reviewer: ```suggestion
<3> The full XML-based logback configuration
```

> . The operator will detect changes on the `ConfigMap` resource, apply

Reviewer: ```suggestion
. The operator detects changes on the `ConfigMap` resource, applies
```

>   the logging configuration and re-deploy all required components.

Reviewer: ```suggestion
  the logging configuration and re-deploys all required components.
```
